File: latest/week3/workshop7_m6/W07_M06_demo.R
# ============================================================
# Workshop 7 — Module 6: Declutter & Redesign
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── LOAD DATA ────────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)), year_added = year(date_added))
yearly <- nf |> filter(!is.na(year_added), year_added >= 2016, year_added <= 2021) |>
  count(year_added, type)
movies <- yearly |> filter(type == "Movie")
tv <- yearly |> filter(type == "TV Show")

# ═══════════════════════════════════════════════════════
# 1. BEFORE: Maximum Clutter (deliberately bad)
# ═══════════════════════════════════════════════════════
p_before <- ggplot(yearly, aes(x = factor(year_added), y = n, fill = type)) +
  geom_col(position = "dodge", width = 0.7, color = "black", linewidth = 0.5) +
  geom_text(aes(label = n), position = position_dodge(0.7),
    vjust = -0.3, size = 2.5, angle = 90) +
  scale_fill_manual(values = c(Movie = "#4472C4", "TV Show" = "#ED7D31")) +
  theme_gray() +
  theme(
    panel.grid.major = element_line(color = "#CCCCCC"),
    panel.grid.minor = element_line(color = "#DDDDDD"),
    panel.border = element_rect(color = "black", fill = NA, linewidth = 2),
    legend.background = element_rect(fill = "white", color = "black"),
    legend.key = element_rect(fill = "white", color = "black"),
    plot.background = element_rect(fill = "#F5F5F5"),
    axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Netflix Additions by Type (2016-2021)",
       subtitle = "Showing the number of titles added per year by content type",
       x = "Year of Addition", y = "Number of Titles Added to the Platform",
       fill = "Content Type",
       caption = "Source: Netflix Dataset from Kaggle (Shivam Bansal, 2021). Data accessed March 2025. All rights reserved.")
ggsave("output/before_cluttered.png", p_before, width = 9, height = 6, dpi = 300)

# ═══════════════════════════════════════════════════════
# 2. AFTER: Decluttered (same data, maximum clarity)
# ═══════════════════════════════════════════════════════
decline <- round((1 - movies$n[movies$year_added == 2021] /
  max(movies$n)) * 100)

p_after <- ggplot(yearly, aes(x = year_added, y = n, group = type)) +
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = movies, color = "#1565C0", linewidth = 2.5) +
  geom_point(data = movies, color = "#1565C0", size = 3) +
  # Callout
  annotate("label", x = 2018.5, y = movies$n[movies$year_added == 2021] * 0.82,
    label = paste0("–", decline, "%\nfrom peak"),
    size = 3.5, fontface = "bold", color = "#E53935",
    fill = "white", label.size = 0.4) +
  # Direct labels
  geom_text(data = yearly |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.1, size = 3.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBB")) +
  coord_cartesian(xlim = c(2016, 2022.5)) +
  theme_minimal(base_size = 11) +
  theme(
    legend.position = "none",
    panel.grid = element_blank(),
    axis.line = element_line(color = "#EEEEEE", linewidth = 0.3),
    plot.title = element_text(face = "bold", size = 13),
    plot.subtitle = element_text(color = "#666", size = 10),
    axis.title.y = element_text(color = "#888")) +
  labs(title = paste0("Movie additions declined ", decline, "% from their 2019 peak"),
       subtitle = "TV Shows grew steadily but Movies — the larger category — contracted",
       x = NULL, y = "Titles Added")
ggsave("output/after_clean.png", p_after, width = 9, height = 5, dpi = 300)

# Side-by-side
ggsave("output/before_after_comparison.png",
  (p_before + labs(title = "BEFORE: 12 clutter elements")) |
  (p_after + labs(title = "AFTER: Decluttered")),
  width = 18, height = 6, dpi = 300)

# ═══════════════════════════════════════════════════════
# 3. THE 12-STEP TRANSFORMATION (annotated)
# ═══════════════════════════════════════════════════════
cat("\n── 12-Step Declutter Transformation ──\n")
steps <- tibble(
  step = 1:12,
  element = c("Chart type", "Title", "Legend", "Borders", "Gridlines", "Data labels",
    "Background", "Axis labels", "Colour", "Bar edges", "Source line", "Spines"),
  before = c("Grouped bar", "Descriptive", "Box with border", "Heavy black",
    "Both axes, grey", "On every bar", "Grey fill", "Verbose", "Two bright",
    "Black outlines", "Long, distracting", "All 4, thick"),
  after = c("Line (continuous)", "Declarative (= finding)", "Direct labels (no legend)",
    "Removed entirely", "Removed entirely", "Only the callout %",
    "White", "Minimal, muted", "Grey + one accent", "None (line chart)",
    "Removed", "Left + bottom, light #EEE"))
print(steps, n = 12)

# ═══════════════════════════════════════════════════════
# 4. CLEAN THEME TEMPLATE
# ═══════════════════════════════════════════════════════
theme_clean <- theme_minimal(base_size = 11) +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(color = "#EEEEEE", linewidth = 0.3),
    plot.title = element_text(face = "bold", size = 14),
    plot.subtitle = element_text(color = "#666", size = 10),
    plot.caption = element_text(color = "#AAA", size = 7),
    legend.background = element_blank(),
    legend.key = element_blank(),
    axis.title = element_text(color = "#888"))

# Apply globally
theme_set(theme_clean)

# Demo: any plot now inherits the clean theme
p_demo <- ggplot(movies, aes(x = year_added, y = n)) +
  geom_line(color = "#1565C0", linewidth = 2) +
  geom_point(color = "#1565C0", size = 3) +
  labs(title = "With theme_set(theme_clean): every plot starts clean",
       subtitle = "No manual theme adjustments needed per chart",
       x = "Year", y = "Movies Added")
ggsave("output/theme_clean_demo.png", p_demo, width = 8, height = 4.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# 5. e-Car MAKEOVER
# ═══════════════════════════════════════════════════════
ec <- read_csv("ecar.csv") |>
  mutate(Year = year(mdy(`Approve Date`)),
    Spread = Rate - `Cost of Funds`) |>
  filter(Year >= 2002, Year <= 2012)
yr_ec <- ec |> group_by(Year) |>
  summarise(rate = mean(Rate), spread = mean(Spread))

# Before: cluttered
p_ec_before <- ggplot(yr_ec) +
  geom_line(aes(x = Year, y = rate, linetype = "Rate"), color = "blue", linewidth = 1) +
  geom_line(aes(x = Year, y = spread, linetype = "Spread"), color = "red", linewidth = 1) +
  geom_point(aes(x = Year, y = rate), color = "blue", size = 3) +
  geom_point(aes(x = Year, y = spread), color = "red", size = 3) +
  scale_linetype_manual(values = c(Rate = "solid", Spread = "dashed")) +
  theme_gray() +
  theme(panel.border = element_rect(color = "black", fill = NA, linewidth = 1.5),
    panel.grid = element_line(color = "#CCC")) +
  labs(title = "e-Car Rate and Spread Over Time",
       x = "Year", y = "Value (%)", linetype = "Metric")

# After: clean
p_ec_after <- ggplot(yr_ec) +
  annotate("rect", xmin = 2007.5, xmax = 2009.5, ymin = -Inf, ymax = Inf,
    fill = "#E53935", alpha = 0.04) +
  geom_line(aes(x = Year, y = spread), color = "#DDDDDD", linewidth = 0.8) +
  geom_line(aes(x = Year, y = rate), color = "#1565C0", linewidth = 2.5) +
  geom_point(aes(x = Year, y = rate), color = "#1565C0", size = 2.5) +
  annotate("text", x = 2012.3, y = yr_ec$rate[yr_ec$Year == 2012],
    label = "Rate", color = "#1565C0", fontface = "bold", size = 3.5, hjust = 0) +
  annotate("text", x = 2012.3, y = yr_ec$spread[yr_ec$Year == 2012],
    label = "Spread", color = "#AAAAAA", size = 3, hjust = 0) +
  annotate("label", x = 2008, y = max(yr_ec$rate) * 0.92,
    label = "2008\ncrisis", size = 2.5, fontface = "bold",
    color = "#E53935", fill = "white", label.size = 0.4) +
  scale_x_continuous(limits = c(2002, 2013)) +
  labs(title = "Loan rates dropped 2pp after the 2008 financial crisis",
       subtitle = "Spread (grey) also compressed; shaded region = crisis window",
       x = NULL, y = "Percentage Points")
ggsave("output/ecar_before_after.png",
  (p_ec_before + labs(title = "BEFORE")) | (p_ec_after + labs(title = "AFTER")),
  width = 16, height = 5, dpi = 300)

# ═══════════════════════════════════════════════════════
# 6. INCREMENTAL DECLUTTER: 4 stages
# ═══════════════════════════════════════════════════════
p_stage0 <- p_before + labs(title = "Stage 0: Original (maximum clutter)")
p_stage1 <- ggplot(yearly, aes(x = factor(year_added), y = n, fill = type)) +
  geom_col(position = "dodge", width = 0.7) +
  scale_fill_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal() + labs(title = "Stage 1: Remove borders, 3D, background", fill = NULL)
p_stage2 <- ggplot(yearly, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1.2) + geom_point(size = 2) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal() + theme(panel.grid = element_blank()) +
  labs(title = "Stage 2: Line chart, remove gridlines", color = NULL)
p_stage3 <- p_after + labs(title = "Stage 3: Grey+accent, direct labels, callout")

ggsave("output/incremental_declutter.png",
  (p_stage0 | p_stage1) / (p_stage2 | p_stage3) +
  plot_annotation(title = "Incremental Declutter: 4 Stages of Transformation"),
  width = 16, height = 10, dpi = 300)

cat("\n── All W07-M06 R outputs saved ──\n")

File: latest/week3/workshop7_m6/W07_M06_demo.py
"""W07-M06: Declutter & Redesign — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True)

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
yearly = nf.query("2016<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type=='Movie'"); tvshows = yearly.query("type=='TV Show'")
peak_n = movies.n.max(); latest_n = movies.iloc[-1]["n"]
decline = round((1 - latest_n / peak_n) * 100)

# ═══════════════════════════════════════════════════════
# 1. BEFORE: Maximum clutter (deliberately bad)
# ═══════════════════════════════════════════════════════
fig, ax = plt.subplots(figsize=(9, 6))
x = np.arange(2016, 2022); w = 0.3
bars1 = ax.bar(x-w/2, movies.n.values, w, color="#4472C4", edgecolor="black", lw=0.5, label="Movie")
bars2 = ax.bar(x+w/2, tvshows.n.values, w, color="#ED7D31", edgecolor="black", lw=0.5, label="TV Show")
# Labels on every bar
for b in list(bars1) + list(bars2):
    ax.text(b.get_x()+b.get_width()/2, b.get_height()+10, f"{b.get_height():.0f}",
        ha="center", fontsize=5, rotation=90)
ax.legend(frameon=True, fancybox=True, shadow=True, fontsize=7)
ax.set_facecolor("#F8F8F8"); ax.grid(True, color="#CCC", lw=0.5)
for spine in ax.spines.values(): spine.set_linewidth(2); spine.set_color("black")
ax.set_title("Netflix Additions by Type (2016-2021)", fontsize=12, fontweight="bold")
ax.set_xlabel("Year of Addition"); ax.set_ylabel("Number of Titles Added to Platform")
ax.tick_params(labelsize=8, rotation=45)
plt.tight_layout(); plt.savefig("output/before_cluttered.png", dpi=300); plt.close()

# ═══════════════════════════════════════════════════════
# 2. AFTER: Decluttered (same data)
# ═══════════════════════════════════════════════════════
fig, ax = plt.subplots(figsize=(9, 5))
ax.plot(tvshows.year_added, tvshows.n, "-o", color="#DDDDDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax.text(2021.15, latest_n, " Movies", fontsize=9, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.15, tvshows.iloc[-1]["n"], " TV Shows", fontsize=8, color="#BBB", va="center")
ax.annotate(f"–{decline}%\nfrom peak", xy=(2019.5, (peak_n+latest_n)/2),
    fontsize=10, fontweight="bold", color="#E53935", ha="center",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935", alpha=0.9))
ax.set_xlim(2015.5, 2022.5)
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.spines["left"].set_color("#EEE"); ax.spines["bottom"].set_color("#EEE")
ax.tick_params(colors="#888", labelsize=9)
ax.set_ylabel("Titles Added", color="#888")
ax.set_title(f"Movie additions declined {decline}% from their 2019 peak",
    fontsize=13, fontweight="bold")
plt.tight_layout(); plt.savefig("output/after_clean.png", dpi=300); plt.close()

# ═══════════════════════════════════════════════════════
# 3. SIDE-BY-SIDE
# ═══════════════════════════════════════════════════════
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(18, 6))

# Before
x_arr = np.arange(2016, 2022); w = 0.3
ax1.bar(x_arr-w/2, movies.n.values, w, color="#4472C4", edgecolor="black", lw=0.5, label="Movie")
ax1.bar(x_arr+w/2, tvshows.n.values, w, color="#ED7D31", edgecolor="black", lw=0.5, label="TV Show")
ax1.legend(frameon=True, shadow=True, fontsize=7)
ax1.set_facecolor("#F8F8F8"); ax1.grid(True, color="#CCC", lw=0.5)
for s in ax1.spines.values(): s.set_linewidth(2); s.set_color("black")
ax1.set_title("BEFORE: 12 clutter elements", fontsize=11, fontweight="bold", color="#C62828")
ax1.set_xlabel("Year"); ax1.set_ylabel("Titles")

# After
ax2.plot(tvshows.year_added, tvshows.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax2.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax2.text(2021.15, latest_n, " Movies", fontsize=9, fontweight="bold", color="#1565C0", va="center")
ax2.text(2021.15, tvshows.iloc[-1]["n"], " TV Shows", fontsize=8, color="#BBB", va="center")
ax2.annotate(f"–{decline}%", xy=(2019.5, (peak_n+latest_n)/2),
    fontsize=11, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935"))
ax2.set_xlim(2015.5, 2022.5)
ax2.spines["top"].set_visible(False); ax2.spines["right"].set_visible(False)
ax2.spines["left"].set_color("#EEE"); ax2.spines["bottom"].set_color("#EEE")
ax2.tick_params(colors="#888"); ax2.set_ylabel("Titles Added", color="#888")
ax2.set_title("AFTER: Decluttered", fontsize=11, fontweight="bold", color="#1565C0")

fig.suptitle("Chart Makeover: Same Data, Radically Different Clarity", fontsize=14, fontweight="bold")
plt.tight_layout(); plt.savefig("output/before_after.png", dpi=300); plt.close()

# ═══════════════════════════════════════════════════════
# 4. CLEAN rcParams TEMPLATE
# ═══════════════════════════════════════════════════════
clean_params = {
    "axes.spines.top": False,
    "axes.spines.right": False,
    "axes.edgecolor": "#EEEEEE",
    "axes.labelcolor": "#888888",
    "xtick.color": "#888888",
    "ytick.color": "#888888",
    "grid.alpha": 0,
    "font.size": 10,
    "axes.titleweight": "bold",
    "figure.facecolor": "white",
    "axes.facecolor": "white",
}
plt.rcParams.update(clean_params)

# Demo with clean defaults
fig, ax = plt.subplots(figsize=(8, 4.5))
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2, markersize=5)
ax.set_title("With rcParams: every plot starts clean")
ax.set_ylabel("Movies Added")
plt.tight_layout(); plt.savefig("output/rcparams_demo.png", dpi=300); plt.close()

# ═══════════════════════════════════════════════════════
# 5. INCREMENTAL DECLUTTER: 4 stages
# ═══════════════════════════════════════════════════════
plt.rcParams.update(plt.rcParamsDefault)  # reset for the "before" stage

fig, axes = plt.subplots(2, 2, figsize=(16, 10))

# Stage 0: original clutter
ax = axes[0,0]
ax.bar(x_arr-w/2, movies.n.values, w, color="#4472C4", edgecolor="black", lw=0.5, label="Movie")
ax.bar(x_arr+w/2, tvshows.n.values, w, color="#ED7D31", edgecolor="black", lw=0.5, label="TV Show")
ax.legend(fontsize=6, frameon=True, shadow=True); ax.grid(True, color="#CCC")
ax.set_facecolor("#F8F8F8"); ax.set_title("Stage 0: Maximum clutter", fontsize=10, fontweight="bold")

# Stage 1: remove borders, background, shadows
ax = axes[0,1]
ax.bar(x_arr-w/2, movies.n.values, w, color="#1565C0", label="Movie")
ax.bar(x_arr+w/2, tvshows.n.values, w, color="#E53935", label="TV Show")
ax.legend(fontsize=6); ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.set_title("Stage 1: Remove borders, background", fontsize=10, fontweight="bold")

# Stage 2: switch to line, remove gridlines
ax = axes[1,0]
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=1.5, label="Movie")
ax.plot(tvshows.year_added, tvshows.n, "-o", color="#E53935", lw=1.5, label="TV Show")
ax.legend(fontsize=6); ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.set_title("Stage 2: Line chart, remove grid", fontsize=10, fontweight="bold")

# Stage 3: grey+accent, direct labels, callout
ax = axes[1,1]
ax.plot(tvshows.year_added, tvshows.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax.text(2021.1, latest_n, " Movies", fontsize=8, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.1, tvshows.iloc[-1]["n"], " TV", fontsize=7, color="#BBB", va="center")
ax.annotate(f"–{decline}%", xy=(2019.5,(peak_n+latest_n)/2),
    fontsize=10, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.2", facecolor="white", edgecolor="#E53935"))
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.spines["left"].set_color("#EEE"); ax.spines["bottom"].set_color("#EEE")
ax.set_xlim(2015.5, 2022); ax.set_title("Stage 3: Grey+accent, labels, callout", fontsize=10, fontweight="bold")

fig.suptitle("Incremental Declutter: 4 Stages", fontsize=14, fontweight="bold")
plt.tight_layout(); plt.savefig("output/incremental_declutter.png", dpi=300); plt.close()

# ═══════════════════════════════════════════════════════
# 6. 12-STEP CHECKLIST (printed)
# ═══════════════════════════════════════════════════════
print("\n=== 12-Step Declutter Transformation ===")
steps = [
    ("Chart type",   "Grouped bar → Line (continuous data)"),
    ("Title",        "Descriptive → Declarative (= finding)"),
    ("Legend",       "Box with border → Direct labels (no legend)"),
    ("Borders",      "Heavy black → Removed entirely"),
    ("Gridlines",    "Both axes, grey → Removed entirely"),
    ("Data labels",  "On every bar → Only the callout %"),
    ("Background",   "Grey fill → White"),
    ("Axis labels",  "Verbose → Minimal, muted colour"),
    ("Colour",       "Two bright → Grey + one accent"),
    ("Bar edges",    "Black outlines → None (line chart)"),
    ("Source line",  "Long, distracting → Removed"),
    ("Spines",       "All 4, thick → Left+bottom, light #EEE"),
]
for i, (element, change) in enumerate(steps, 1):
    print(f"  {i:2d}. {element:<15} {change}")

print("\nAll W07-M06 Python outputs saved")
